azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -526,3 +526,184 @@ def _handle_pipeline_event(self, event):
                 )
         else:
             self._send_event_up(event)
+
+
+class TimeoutStage(PipelineStage):
+    """
+    The purpose of the timeout stage is to add timeout errors to select operations
+
+    The timeout_intervals attribute contains a list of operations to track along with
+    their timeout values.  Right now this list is hard-coded but the operations and
+    intervals will eventually become a parameter.
+
+    For each operation that needs a timeout check, this stage will add a timer to
+    the operation.  If the timer elapses, this stage will fail the operation with
+    a PipelineTimeoutError.  The intention is that a higher stage will know what to
+    do with that error and act accordingly (either return the error to the user or
+    retry).
+
+    This stage currently assumes that all timed out operatoins are just "lost".
+    It does not attempt to cancel the operation, as Paho doesn't have a way to
+    cancel an operation, and with QOS=1, sending a pub or sub twice is not
+    catastrophic.
+
+    The long term intention of this stage is that the timeout intervals themselves
+    will eventually come from a retry policy.
+
+    Also, as a long-term plan, the operations that need to be watched for timeout
+    will become an initialization parameter for this stage so that differet
+    instances of this stage can watch for timeouts on different operations.
+    This will be done because we want a lower-level timeout stage which can watch
+    for timeouts at the MQTT level, and we want a higher- level timeout stage which
+    can watch for timeouts at the iothub level.  In this way, an MQTT operation that
+    times out can be retried as an MQTT operation and a higher-level IoTHub operation
+    which times out can be retried as an IoTHub operation (which might necessitate
+    redoing multiple MQTT operations).
+    """
+
+    def __init__(self):
+        super(TimeoutStage, self).__init__()
+        self.timeout_intervals = {
+            pipeline_ops_mqtt.MQTTSubscribeOperation: 10,
+            pipeline_ops_mqtt.MQTTUnsubscribeOperation: 10,
+        }
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _should_watch_for_timeout(self, op):
```

Comment:
>_should_watch_for_timeout [](start = 8, length = 25)

since this is when op needs to have timer m,ay be it should be renamed to `should_have_timer` or `should_be_timed_via_timer`  ?